# Assignment 1
### Robert Kraemer 
<b>Due:</b> by midnight MST the day prior to Live Session 2 

In [230]:
import csv
import random
import numpy as np
import pandas as pd

### Assignment 1, Part 1: 
Construct a CSV file with the first eight elements of the periodic table. Include columns for name, symbol, and atomic number. Read that into a pandas DataFrame. Inside the program, add a ninth and 10th element, and then add a column with the atomic weights rounded to the nearest integer.

In [231]:
# Construct a CSV file with the first eight elements of the periodic table
periodic_header = ['name', 'symbol', 'atomic_number']
periodic_data = [['Hydrogen','H',1],
                 ['Helium','He',2], 
                 ['Lithium','Li',3],
                 ['Beryllium','Be',4],
                 ['Boron','B',5],
                 ['Carbon','C',6],
                 ['Nitrogen','N',7],
                 ['Oxygen','O',8]
                ]
with open('periodic_table.csv', 'w') as file:
    writer = csv.writer(file)
    writer.writerow(periodic_header)
    writer.writerows(periodic_data)

In [232]:
# Read CSV file into a pandas DataFrame
periodic_df = pd.read_csv('periodic_table.csv')
periodic_df

name symbol  atomic_number
0   Hydrogen      H              1
1     Helium     He              2
2    Lithium     Li              3
3  Beryllium     Be              4
4      Boron      B              5
5     Carbon      C              6
6   Nitrogen      N              7
7     Oxygen      O              8

In [233]:
# add a 9th and 10th element
nine_and_ten = pd.DataFrame({'name':['Fluorine','Neon'],
                             'symbol':['F','Ne'],
                             'atomic_number':[9,10]})
periodic_df = pd.concat([periodic_df, nine_and_ten], ignore_index = True).reset_index(drop=True)
periodic_df

name symbol  atomic_number
0   Hydrogen      H              1
1     Helium     He              2
2    Lithium     Li              3
3  Beryllium     Be              4
4      Boron      B              5
5     Carbon      C              6
6   Nitrogen      N              7
7     Oxygen      O              8
8   Fluorine      F              9
9       Neon     Ne             10

In [234]:
# add a column with the atomic weights rounded to the nearest integer
periodic_df['atomic_weights'] = [1,4,7,9,11,12,14,16,19,20]
periodic_df

name symbol  atomic_number  atomic_weights
0   Hydrogen      H              1               1
1     Helium     He              2               4
2    Lithium     Li              3               7
3  Beryllium     Be              4               9
4      Boron      B              5              11
5     Carbon      C              6              12
6   Nitrogen      N              7              14
7     Oxygen      O              8              16
8   Fluorine      F              9              19
9       Neon     Ne             10              20

### Assignment 1, Part 2: 
Make a list of strings for nine Greek letters, for example ‘alpha’. Make that list such that the strings are not in alphabetic order. Make two 9-element numpy arrays of random floating-point numbers with an estimated mean of 10 and a standard deviation of 1.5. Make another array of nine elements ranging from zero to two times pi. Name it ‘angle’. Make another array holding the cosine of that ‘angle’ array. Construct a dictionary from all of the above. Form a DataFrame from that dictionary, and print it out. Sort the DataFrame ascending on the Greek letters, drop two columns of your choice, drop one of the rows, and print that out.

In [235]:
# Make a list of strings for nine Greek letters, for example ‘alpha’
# Make that list such that the strings are not in alphabetic order
greek_letters = ['alpha','beta','gamma','delta','epsilon','zeta','eta','theta','iota']
print('\nGreek letters:', greek_letters)


Greek letters: ['alpha', 'beta', 'gamma', 'delta', 'epsilon', 'zeta', 'eta', 'theta', 'iota']


In [236]:
# Make two 9-element numpy arrays of random floating-point numbers with an 
# estimated mean of 10 and a standard deviation of 1.5
array1 = np.random.normal(loc=10, scale=1.5, size=9)
print('\nArray 1:', array1) 
array2 = np.random.normal(loc=10, scale=1.5, size=9)
print('\nArray 2:', array2) 


Array 1: [ 9.63653469 11.33567704  9.67532703  8.06501194 10.15506943 10.45224344
  9.58671532  8.5571646  11.29004329]

Array 2: [10.68730024 10.02392885 11.12453476  7.96781164  9.56365836 11.20441546
 11.05050578  9.92022541  9.6230921 ]


In [237]:
# Make another array of nine elements ranging from zero to two times pi
# Name it ‘angle’
angle = np.random.uniform(low=0, high=2*np.pi, size=9)
print('\nAngle:', angle)


Angle: [5.53751415 3.72030843 2.24826571 0.7012673  0.20123495 0.72955824
 5.76353377 2.70030426 6.00621121]


In [238]:
# Make another array holding the cosine of that ‘angle’ array
cos_angle = np.cos(angle)
print('\nCosine of ‘angle’ array:', cos_angle)


Cosine of ‘angle’ array: [ 0.73463271 -0.83716574 -0.62682328  0.76402516  0.97982048  0.74546893
  0.86799227 -0.90420214  0.96188726]


In [239]:
# Construct a dictionary from all of the above
dict_container = {'greek_letters':greek_letters,
                  'array1':array1,
                  'array2':array2,
                  'angle':angle,
                  'cos_angle':cos_angle
                 }
dict_container

{'greek_letters': ['alpha',
  'beta',
  'gamma',
  'delta',
  'epsilon',
  'zeta',
  'eta',
  'theta',
  'iota'],
 'array1': array([ 9.63653469, 11.33567704,  9.67532703,  8.06501194, 10.15506943,
        10.45224344,  9.58671532,  8.5571646 , 11.29004329]),
 'array2': array([10.68730024, 10.02392885, 11.12453476,  7.96781164,  9.56365836,
        11.20441546, 11.05050578,  9.92022541,  9.6230921 ]),
 'angle': array([5.53751415, 3.72030843, 2.24826571, 0.7012673 , 0.20123495,
        0.72955824, 5.76353377, 2.70030426, 6.00621121]),
 'cos_angle': array([ 0.73463271, -0.83716574, -0.62682328,  0.76402516,  0.97982048,
         0.74546893,  0.86799227, -0.90420214,  0.96188726])}

In [240]:
# Form a DataFrame from that dictionary, and print it out
df = pd.DataFrame(dict_container)
df

greek_letters     array1     array2     angle  cos_angle
0         alpha   9.636535  10.687300  5.537514   0.734633
1          beta  11.335677  10.023929  3.720308  -0.837166
2         gamma   9.675327  11.124535  2.248266  -0.626823
3         delta   8.065012   7.967812  0.701267   0.764025
4       epsilon  10.155069   9.563658  0.201235   0.979820
5          zeta  10.452243  11.204415  0.729558   0.745469
6           eta   9.586715  11.050506  5.763534   0.867992
7         theta   8.557165   9.920225  2.700304  -0.904202
8          iota  11.290043   9.623092  6.006211   0.961887

In [241]:
# Sort the DataFrame ascending on the Greek letters, 
# drop two columns of your choice, drop one of the rows, 
# and print that out.
df_update = df.sort_values(['greek_letters'],ascending=True)\
                .drop(['array2','cos_angle'],axis=1)\
                .drop(index=4,axis=0)\
                .reset_index(drop=True)
df_update

greek_letters     array1     angle
0         alpha   9.636535  5.537514
1          beta  11.335677  3.720308
2         delta   8.065012  0.701267
3           eta   9.586715  5.763534
4         gamma   9.675327  2.248266
5          iota  11.290043  6.006211
6         theta   8.557165  2.700304
7          zeta  10.452243  0.729558

### Assignment 1, Part 3: 
Write a program in Python to create and print out the first 12 Fibonacci numbers. Then iterate over the last five numbers to build another list with the ratio of each number to its predecessor. What do you observe about this latter list?

In [242]:
def fib(n=11):
    if n<2:
        return n
    fibs = [0,1]
    for i in range(2, n+1):
        next = fibs[-1]+fibs[-2]
        fibs.append(next)
    print('Fibonacci numbers:', fibs)
    
    fib_ratio=[]
    for i in range(5): 
        last = fibs.pop()
        pred = fibs[-1]
        fib_ratio.append(last/pred)
        #print(fibs)
    print('Ratio:', fib_ratio)

fib()

Fibonacci numbers: [0, 1, 1, 2, 3, 5, 8, 13, 21, 34, 55, 89]
Ratio: [1.6181818181818182, 1.6176470588235294, 1.619047619047619, 1.6153846153846154, 1.625]


The quotient between each successive pair of Fibonacci numbers approximates 1.618, or its inverse 0.618

### Assignment 1, Part 4: 
Provide a function that converts temperature in Kelvin to Rankine. Make a list of five Kelvin temperatures, and print out their values in Rankine. Repeat using a lambda function.

In [243]:
# Provide a function that converts temperature in Kelvin to Rankine 
def kelvin_to_rankine(k): 
    r = 1.8*k
    return r
kelvin_to_rankine(100)

180.0

In [244]:
# Make a list of five Kelvin temperatures, and print out their values in Rankine
k_temps = [100,150,200,250,300]
for tmp in k_temps:
    print(kelvin_to_rankine(tmp))

180.0
270.0
360.0
450.0
540.0


In [245]:
# Repeat using a lambda function
conversion = lambda x: 1.8*x
print(*list(map(conversion, k_temps)),sep='\n')

180.0
270.0
360.0
450.0
540.0
